skimage/registration/_lddmm.py

```diff
@@ -0,0 +1,2200 @@
+            # to 1 and avoid division by 0.
+            velocity_fields_update_norm[velocity_fields_update_norm == 0] = 1
+            velocity_fields_update = (
+                velocity_fields_update
+                / velocity_fields_update_norm[..., None]
+                * np.arctan(
+                    velocity_fields_update_norm[..., None]
+                    * np.pi
+                    / 2
+                    / self.maximum_velocity_fields_update
+                )
+                * self.maximum_velocity_fields_update
+                * 2
+                / np.pi
+            )
+
+            self.velocity_fields[..., timestep, :] -= velocity_fields_update
+
+        # Compute and save maximum velocity for diagnostics.
+        maximum_velocity = np.sqrt(
+            np.sum(self.velocity_fields ** 2, axis=-1)
+        ).max()
+        self.maximum_velocities.append(maximum_velocity)
+
+    def _compute_affine_phi(self):
+        """
+        Compute and set self.affine_phi. Called once in case there were no
+        deformative iterations to set it.
+
+        Accesses attributes:
+            reference_image_axes
+            reference_image_coords
+            delta_t
+            num_timesteps
+            velocity_fields
+            affine
+            phi
+            affine_phi
+
+        Updates attributes:
+            phi
+            affine_phi
+        """
+
+        # Set self.phi to identity. self.phi is secretly phi_1t_inv but at the
+        # end of the loop it will be phi_10_inv = phi_01 = phi.
+        self.phi = np.copy(self.reference_image_coords)
+
+        # Loop backwards across time.
+        for timestep in reversed(range(self.num_timesteps)):
+
+            # Update phi.
+            sample_coords = (
+                self.reference_image_coords
+                + self.velocity_fields[..., timestep, :] * self.delta_t
+            )
+            self.phi = (
+                interpn(
+                    points=self.reference_image_axes,
+                    values=self.phi - self.reference_image_coords,
+                    xi=sample_coords,
+                    bounds_error=False,
+                    fill_value=None,
+                )
+                + sample_coords
+            )
+
+            # Apply affine by multiplication.
+            # This transforms error in the moving_image space back to time t.
+            self.affine_phi = _multiply_coords_by_affine(self.affine, self.phi)
+
+    # End _Lddmm.
+
+
+def lddmm_register(
+    reference_image,
+    moving_image,
+    reference_image_spacing=None,
+    moving_image_spacing=None,
+    deformative_stepsize=None,
+    sigma_regularization=None,
+    contrast_order=None,
+    **kwargs,
+):
+    """
+    Compute a registration between grayscale images reference_image and
+    moving_image, to be applied with scipy.ndimage.map_coordinates.
+
+    Parameters
+    ----------
+        reference_image: np.ndarray
```

Comment:
@DevinCrowley These should not be indented. Additionally, you later have additional levels of indentation. I think that's what's making sphinx unhappy. You might try to write it out without nested indentation.

Also, note that you can build the docs locally, see here:

https://scikit-image.org/docs/dev/contribute.html#building-docs

That should make it marginally less painful to work with Sphinx 😂

Reply:
_Blesssss_

Reply:
Can't believe I missed this, thanks for putting some fresh eyes on it!